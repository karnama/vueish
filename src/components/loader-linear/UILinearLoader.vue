<template>
    <div class="w-full overflow-hidden ui-linear-loader"
         :class="inheritColor ? '' : 'text-brand-600'"
         :style="{ 'height': height + 'px' }">
        <div class="relative h-full">
            <div class="progress-background w-full absolute h-full" />
            <div class="absolute h-full"
                 role="progressbar"
                 v-bind="ariaAttributes"
                 :class="determinate ? 'bg-current' : 'indeterminate'"
                 :style="{ width }" />
        </div>
    </div>
</template>

<script lang="ts">
import { defineComponent, computed } from 'vue';
import { inheritColor, progress, steps, determinate } from 'composables/style';

export default defineComponent({
    name: 'UILinearLoader',

    props: {
        /**
         * The height of the component in px.
         */
        height: {
            type: Number,
            default: 4,
            validator: (value: number) => value > 0
        },

        inheritColor,
        determinate,
        progress,
        steps
    },

    setup(props) {
        if (props.determinate) {
            if (!props.steps) {
                throw new Error(
                    'UILinearLoader - The prop is set to finite loading without giving a number of steps'
                );
            }
            if (props.progress! > Number(props.steps)) {
                throw new Error('UILinearLoader  - The progress cannot be higher that the steps');
            }
        }

        const width = computed(() => {
            if (!props.determinate) {
                return '100%';
            }

            return String(100 / Number(props.steps) * props.progress!) + '%';
        });
        const ariaAttributes = computed(() => {
            if (!props.determinate) {
                return {};
            }

            return {
                ariaValueNow: props.progress,
                ariaValueMin: 0,
                ariaValueMax: props.steps
            };
        });

        return {
            width,
            ariaAttributes
        };
    }
});
</script>

<style lang="scss" scoped>
.progress-background {
    background-clip: padding-box;
    background-color: currentColor;
    opacity: 0.2;
}

.indeterminate:before {
    content: '';
    z-index: 1000;
    position: absolute;
    background-color: currentColor;
    filter: brightness(110%);
    top: 0;
    left: 0;
    bottom: 0;
    will-change: left, width;
    animation: increase 2s infinite;
}

.indeterminate:after {
    content: '';
    z-index: 1000;
    position: absolute;
    background-color: currentColor;
    filter: brightness(110%);
    opacity: 1;
    top: 0;
    left: 0;
    bottom: 0;
    will-change: left, width;
    animation: decrease 2s 0.5s infinite;
}

@keyframes increase {
    from {
        left: -5%;
        width: 5%;
    }
    to {
        left: 130%;
        width: 100%;
    }
}

@keyframes decrease {
    from {
        left: -80%;
        width: 80%;
    }
    to {
        left: 110%;
        width: 10%;
    }
}
</style>
