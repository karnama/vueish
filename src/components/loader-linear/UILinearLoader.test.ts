import UILinearLoader from './UILinearLoader.vue';
import { mount } from '@vue/test-utils';

const wrapper = mount(UILinearLoader, {
    props: {
        steps: 10,
        progress: 1,
        height: 5
    }
});

describe('UILinearLoader', () => {
    it('should render', () => {
        expect(wrapper).toBeDefined();
        expect(wrapper.exists()).toBe(true);
    });

    it('should set the height of the element by props', () => {
        expect(getComputedStyle(wrapper.element).height).toBe('5px');
    });

    it('should calculate the correct width', async () => {
        await wrapper.setProps({ determinate: true });
        expect(wrapper.vm.width).toBe('10%');
    });

    it('should throw and error on logical errors in the passed props', () => {
        const consoleErrorMock = jest.spyOn(console, 'warn').mockImplementation(() => {});

        let failingFunc = () => {
            mount(UILinearLoader, {
                props: {
                    determinate: true
                }
            });
        };
        expect(failingFunc).toThrow(new TypeError(
            'UILinearLoader - The prop is set to finite loading without giving a number of steps'
        ));

        failingFunc = () => {
            mount(UILinearLoader, {
                props: {
                    determinate: true,
                    steps: 10,
                    progress: 11
                }
            });
        };
        expect(failingFunc).toThrow(new TypeError('UILinearLoader  - The progress cannot be higher that the steps'));

        consoleErrorMock.mockRestore();
    });
});
