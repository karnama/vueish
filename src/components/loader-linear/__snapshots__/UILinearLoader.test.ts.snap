// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UILinearLoader should render correctly 1`] = `
<div
  class="w-full overflow-hidden ui-linear-loader text-brand-500"
  style="height: 5px;"
>
  <div
    class="relative h-full"
  >
    <div
      class="progress-background w-full absolute h-full"
    />
    <div
      class="absolute h-full indeterminate"
      role="progressbar"
      style="width: 100%;"
    />
  </div>
</div>
`;

exports[`UILinearLoader should render correctly when set to determinate 1`] = `
<div
  class="w-full overflow-hidden ui-linear-loader text-brand-500"
  style="height: 5px;"
>
  <div
    class="relative h-full"
  >
    <div
      class="progress-background w-full absolute h-full"
    />
    <div
      ariavaluemax="10"
      ariavaluemin="0"
      ariavaluenow="1"
      class="absolute h-full bg-current"
      role="progressbar"
      style="width: 10%;"
    />
  </div>
</div>
`;
